src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -65,6 +55,23 @@ internal abstract partial class AbstractProject : ForegroundThreadAffinitizedObj
 
         private bool _pushingChangesToWorkspaceHosts;
 
+        #endregion
+
+        #region Mutable fields accessed only from the foreground - does not need locking for access.
+        /// <summary>
+        /// When a reference changes on disk we start a delayed task to update the <see cref="Workspace"/>.
+        /// It is delayed for two reasons: first, there are often a bunch of change notifications in quick succession
+        /// as the file is written.  Second, we often get the first notification while something is still writing the
+        /// file, so we're unable to actually load it.  To avoid both of these issues, we wait five seconds before
+        /// reloading the metadata.  This <see cref="Dictionary{TKey, TValue}"/> holds on to
+        /// <see cref="CancellationTokenSource"/>s that allow us to cancel the existing reload task if another file
+        /// change comes in before we process it.
+        /// </summary>
+        private readonly Dictionary<VisualStudioMetadataReference, CancellationTokenSource> _changedReferencesPendingUpdate
+            = new Dictionary<VisualStudioMetadataReference, CancellationTokenSource>();
```

Comment:
Consider: 

1. Ban all uses of the field directly 
2. Add a `private` property to access the field
3. The `private` property `get` should assert it is on the foreground thread